Update AWS roles ARNs displayed on `tsh app login` for AWS console apps

gabrielcorado opened the pull request:
Closes https://github.com/gravitational/customer-sensitive-requests/issues/303 and https://github.com/gravitational/teleport/issues/44410.

Update `tsh` to display AWS roles ARNs from root and leaf clusters (when applicable) instead of returning the roles from the profile. This update matches the results shown on Web UI (updated by #44611).

The flow is the same used by the API endpoint called by the Web UI:
- Fetch the applications using `GetEnrichedResourcesPage`, and use the `Logins` field to present the available roles ARNs.
- If the `Logins` field is empty (due to an older version of the server), fallback to retrieving the logins using the access checker `GetAllowedLoginsForResource`. This fallback can be removed in later versions (the same as the Web UI changes).

changelog: Correctly display available allowed logins of leaf AWS Console Apps on `tsh app login`.

gabrielcorado committed:
feat(tsh): list aws console logins from server

gabrielcorado committed:
chore(services): remove unified resources change

This is being covered on another PR.

greedy52 reviewed (commented):
Overall LGTM and tested 🎉 . Left a few comments.

gabrielcorado commented on lib/client/api.go:
```diff
@@ -2550,6 +2550,82 @@ func (tc *TeleportClient) ListApps(ctx context.Context, customFilter *proto.List
+				continue
+			}
+
+			apps = append(apps, appWithLogins{app: appServer.GetApp(), logins: r.Logins})
```
The naming needs to be corrected here. It should be `getAppsWithLogins`. Initially, I updated other app listings to use this function, but the PR scope became too large, so I reduced it to only being used on the new function. If/When we revisit this code, we can update other parts to reuse this function so we only use a single source of the app listing.

gabrielcorado commented on lib/client/api.go:
```diff
@@ -2550,6 +2550,82 @@ func (tc *TeleportClient) ListApps(ctx context.Context, customFilter *proto.List
+
+	var apps []appWithLogins
+	for {
+		page, next, err := client.GetUnifiedResourcePage(ctx, clt, &req)
```
Yes, I agree. It better fits what we're doing here. I've changed it because it wasn't returning the logins properly (due to a bug that will be resolved by #44992). Once we get that merged, I'll update this.

gabrielcorado commented on lib/client/api_test.go:
```diff
@@ -1334,3 +1364,74 @@ func TestGetTargetNodes(t *testing.T) {
 	}
 }
+
+func TestGetAppsWithLogin(t *testing.T) {
```
The naming of the test is correct. [More info on my other comment.](https://github.com/gravitational/teleport/pull/44983#discussion_r1702185674)

gabrielcorado commented on tool/tsh/common/app.go:
```diff
@@ -518,11 +518,15 @@ func getAppInfo(cf *CLIConf, tc *client.TeleportClient, matchRouteToApp func(tls
 		return nil, trace.Wrap(err)
 	}
 
+	if len(logins) == 0 {
+		logins = getARNFromRoles(cf, tc, profile, app)
+	}
```
> should this also be deleted in 18.0?

The idea is that once we delete the `getARNFromRoles`, we also drop this if statement. If that needs to be clarified, I can also add the deletion note here.

> len(logins) == 0 could be a legit scenario where the user just doesn't have any AWS roles assigned. Is it possible to differentiate the two different scenarios?

No. The `Login` field is already on the resource and wasn't being filled out for apps. But, if that happens, the final result will be the same (because the access checker will return empty).

greedy52 replied:
> But, if that happens, the final result will be the same (because the access checker will return empty).

Extra API alls though. But i think it would rare.

gabrielcorado committed:
test(tsh): solve TestAzure flakiness by waiting using app servers are ready

gabrielcorado committed:
fix(tsh): apps with logins were fallingback into using aws arns

gabrielcorado committed:
refactor(client): use GetEnrichedResources

gabrielcorado committed:
chore(client): rename function

gabrielcorado commented on lib/client/api.go:
```diff
@@ -2550,6 +2550,82 @@ func (tc *TeleportClient) ListApps(ctx context.Context, customFilter *proto.List
 	return types.DeduplicateApps(apps), nil
 }
 
+// GetAppWithLogins returns an application with available logins.
+func (tc *TeleportClient) GetAppWithLogins(ctx context.Context, predicateExpression string) (types.Application, []string, error) {
+	ctx, span := tc.Tracer.Start(
+		ctx,
+		"teleportClient/GetAppWithLogins",
+		oteltrace.WithSpanKind(oteltrace.SpanKindClient),
+		oteltrace.WithAttributes(attribute.String("predicate", predicateExpression)),
+	)
+	defer span.End()
+
+	clt, err := tc.ConnectToCluster(ctx)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	defer clt.Close()
```
I've talked with @greedy52, and we think it is better to move the `GetEnrichedResourcePage` to the `tsh` package (similar to what is done for database commands). So we don't need this additional function on `TeleportClient` and can deal better with the reuse of cluster client.

greedy52 commented on tool/tsh/common/app.go:
```diff
@@ -518,11 +518,15 @@ func getAppInfo(cf *CLIConf, tc *client.TeleportClient, matchRouteToApp func(tls
 
 	// If we didn't find an active profile for the app, get info from server.
-	app, err := getApp(cf.Context, tc, cf.AppName)
+	app, logins, err := getApp(cf.Context, tc, cf.AppName)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	if len(logins) == 0 && app.IsAWSConsole() {
+		logins = getARNFromRoles(cf, tc, profile, app)
```
alternatively just fallback to `profile.AWSRolesARNs` without creating the access checker. normally ppl should update their server first before updating clients.

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins

gabrielcorado committed:
refactor(tsh): directly resource lisiting for apps and reuse cluster client

gabrielcorado committed:
chore(client): reset client changes

gabrielcorado committed:
refactor(tsh): reuse cluster client for fetching allowed logins

gabrielcorado committed:
chore(tsh): remove unused function param

gabrielcorado commented:
@greedy52 @rosstimothy I've updated the PR to list the apps using `api/client.GetEnrichedResourcePage` directly (similar to DB commands). Also, I've updated the `getAppInfo` to return the used cluster client so other functions can reuse it instead of creating multiple connections.

gabrielcorado commented on tool/tsh/common/app.go:
```diff
@@ -49,12 +50,17 @@ func onAppLogin(cf *CLIConf) error {
 	}
 
-	appInfo, err := getAppInfo(cf, tc, nil /*matchRouteToApp*/)
+	appInfo, clusterClient, err := getAppInfo(cf, tc, nil /*matchRouteToApp*/)
```
I've placed it because we can reuse the `RetryWithRelogin` (which is always required on `getAppInfo` to fetch the profiles). Also, this is the first function executed by the app commands. It is convenient instead of creating the cluster client outside for each app command.

gabrielcorado replied:
After some testing, I noticed this update caused some cases where login is required (For example, if the device trust configurations change). For those cases, the `ConnectToCluster` operation will succeed (given it just creates the gRPC clients and doesn't immediately connect), but the `GetEnrichedResourcePage` will fail. That's why the previous implementation was called `tc.ListApps` inside a `tc.RetryWithLogin`.

Clients created with `ConnectToCluster` will work as expected because their TLS configuration is always fetched from the `TeleportClient`.

So those functions that perform the app listing should still receive the `TeleportClient` + do the listing within `RetryWithLogin`. Alternatively, we could create the cluster client and perform a gRPC call to initialize the connection.

gabrielcorado replied:
That said, we could retrieve the profile (within login retry) and the cluster client outside it (inside a new function or not). The `getAppInfo` would take both as arguments, as well as the `TeleportClient`, so we can perform the listing with login retry.

gabrielcorado replied:
I've updated the `getAppInfo` to receive `TeleportClient` (which is used later for fetching the app with login retry) and auth client.

greedy52 commented on tool/tsh/common/app.go:
```diff
@@ -617,23 +623,30 @@ func (a *appInfo) GetApp(ctx context.Context, tc *client.TeleportClient) (types.
+		ResourceType:        types.KindAppServer,
+		SortBy:              types.SortBy{Field: types.ResourceMetadataName},
+		PredicateExpression: fmt.Sprintf(`name == "%s"`, name),
+		Limit:               1,
```
In HA setup, you can have multiple KindAppServer for the same app?

gabrielcorado committed:
refactor(tsh): update getApp retry with login

greedy52 commented on tool/tsh/common/app.go:
```diff
@@ -617,23 +622,38 @@ func (a *appInfo) GetApp(ctx context.Context, tc *client.TeleportClient) (types.
-	err = client.RetryWithRelogin(ctx, tc, func() error {
-		apps, err = tc.ListApps(ctx, &proto.ListResourcesRequest{
-			Namespace:           tc.Namespace,
+func getApp(ctx context.Context, tc *client.TeleportClient, clt apiclient.GetResourcesClient, name string) (app types.Application, logins []string, err error) {
```
nit: maybe add a trace?

greedy52 commented on tool/tsh/common/app.go:
```diff
@@ -617,23 +622,38 @@ func (a *appInfo) GetApp(ctx context.Context, tc *client.TeleportClient) (types.
 			ResourceType:        types.KindAppServer,
+			SortBy:              types.SortBy{Field: types.ResourceMetadataName},
 			PredicateExpression: fmt.Sprintf(`name == "%s"`, name),
+			Limit:               1,
```
Is the `PredicateExpression` by the backend or is it used as a filter after the backend ranges?

If later, is limit 1 used in each backend call?

gabrielcorado commented on tool/tsh/common/app.go:
```diff
@@ -49,31 +50,32 @@ func onAppLogin(cf *CLIConf) error {
 	}
 
-	appInfo, err := getAppInfo(cf, tc, nil /*matchRouteToApp*/)
+	clusterClient, err := tc.ConnectToCluster(cf.Context)
```
It won't return any error if a login is required (because it doesn't connect to the cluster, it only creates the client).

greedy52 replied:
Here is what I get though:
```
$ tsh status
> Profile URL:        https://steve-xin.teleport.sh:443
  Logged in as:       xin.huang@goteleport.com
  Cluster:            steve-xin.teleport.sh
  Roles:              access, auditor, editor
  Logins:             *, oracle, root, opc, grid
  Kubernetes:         enabled
  Valid until:        2024-08-12 21:41:06 -0400 EDT [EXPIRED]
  Extensions:         login-ip, permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy

ERROR: Active profile expired.

$ tsh apps login some-app
ERROR: ssh: cert has expired

$ tsh apps login some-app --debug
2024-08-14T10:54:18-04:00 INFO [CLIENT]    ALPN connection upgrade required for "steve-xin.teleport.sh:443": false. client/api.go:838
2024-08-14T10:54:18-04:00 INFO [CLIENT]    no host login given. defaulting to stevehuang client/api.go:1181
2024-08-14T10:54:18-04:00 INFO [CLIENT]    [KEY AGENT] Connected to the system agent: "/private/tmp/com.apple.launchd.LmAabEuuQ8/Listeners" client/api.go:4586
2024-08-14T10:54:18-04:00 DEBU [KEYSTORE]  Teleport TLS certificate valid until "2024-08-13 01:41:06 +0000 UTC". client/client_store.go:118
2024-08-14T10:54:18-04:00 INFO [KEYAGENT]  Loading SSH key for user "xin.huang@goteleport.com" and cluster "steve-xin.teleport.sh". client/keyagent.go:198
2024-08-14T10:54:18-04:00 DEBU [KEYSTORE]  Teleport TLS certificate valid until "2024-08-13 01:41:06 +0000 UTC". client/client_store.go:118

ERROR REPORT:
Original Error: *errors.errorString ssh: cert has expired
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/utils/sshutils/checker.go:68 github.com/gravitational/teleport/api/utils/sshutils.(*CertChecker).CheckCert
	github.com/gravitational/teleport/lib/client/interfaces.go:618 github.com/gravitational/teleport/lib/client.(*KeyRing).checkCert
	github.com/gravitational/teleport/lib/client/keyagent.go:627 github.com/gravitational/teleport/lib/client.(*LocalKeyAgent).Signers
	github.com/gravitational/teleport/lib/client/api.go:3020 github.com/gravitational/teleport/lib/client.(*TeleportClient).generateClientConfig
	github.com/gravitational/teleport/lib/client/api.go:2909 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToCluster
	github.com/gravitational/teleport/tool/tsh/common/app.go:53 github.com/gravitational/teleport/tool/tsh/common.onAppLogin
	github.com/gravitational/teleport/tool/tsh/common/tsh.go:1459 github.com/gravitational/teleport/tool/tsh/common.Run
	github.com/gravitational/teleport/tool/tsh/common/tsh.go:608 github.com/gravitational/teleport/tool/tsh/common.Main
	github.com/gravitational/teleport/tool/tsh/main.go:26 main.main
	runtime/proc.go:271 runtime.main
	runtime/asm_arm64.s:1222 runtime.goexit
User Message: ssh: cert has expired
```

gabrielcorado replied:
Done. I've updated it to use a single function to grab the profile and create the cluster client.

gabrielcorado committed:
refactor(tsh): use a single function to grab profile and cluste client

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins

greedy52 commented on tool/tsh/common/app.go:
```diff
@@ -49,31 +50,32 @@ func onAppLogin(cf *CLIConf) error {
 	}
 
-	appInfo, err := getAppInfo(cf, tc, nil /*matchRouteToApp*/)
+	clusterClient, profile, err := initClusterAndProfile(cf.Context, tc)
```
close the client?

greedy52 commented on tool/tsh/common/app.go:
```diff
@@ -470,6 +471,25 @@ func formatAppConfig(tc *client.TeleportClient, profile *client.ProfileStatus, r
 	}
 }
 
+// initClusterAndProfile initializes a cluster client and also returns the
```
`initClusterAndProfile` sounds like a factory method that creates a cluster. maybe we can improve the name a little

gabrielcorado commented on tool/tsh/common/app.go:
```diff
@@ -617,23 +632,38 @@ func (a *appInfo) GetApp(ctx context.Context, tc *client.TeleportClient) (types.
+	// login is required (e.g., configuration change). We don't need a new
+	// instance of GetResourcesClient since it will reconnect using credentials
+	// from the updated TeleportClient.
+	if err := client.RetryWithRelogin(ctx, tc, func() error {
```
I've moved all interactions that require to be on a retry with login into a single block.

gabrielcorado committed:
refactor(tsh): perform retry with login at caller site

gabrielcorado committed:
fix(tsh): close auth client

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins

gabrielcorado commented:
I'll update it later today

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins

gabrielcorado committed:
test(tsh): fix test failing due to login misconfiguration

gabrielcorado commented:
@rosstimothy Tests should be fixed now. The flaky detector is failing due to timeout, and we might need to exclude `TestAWSConsoleLogins` (which takes 16s on average to complete)

gabrielcorado commented:
> Is there anything that we can do to make TestAWSConsoleLogins faster?

All three of my modified tests (`TestAzure`, `TestAWS`, and `TestAWSConsoleLogins`) are timing out. Most of the time spent on those tests is setting up the clusters.

As for `TestAWSConsoleLogins`, we could split it into two parts: one for the root cluster (which will require a single cluster) and another for the leaf (which requires root and leaf cluster). This split would not guarantee the first test to pass (on my machine, it took, on average, 11s). The second test still averages (15s) to complete, so it will most likely time out.

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins

gabrielcorado committed:
test(tsh): fix lint errors

gabrielcorado committed:
Merge branch 'gabrielcorado/tsh-app-login-leaf-logins' of github.com:gravitational/teleport into gabrielcorado/tsh-app-login-leaf-logins

gabrielcorado committed:
test(tsh): remove unused imports

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tsh-app-login-leaf-logins
